### ISSUE TITLE
Cannot attach to running instances when global process hook

### ISSUE THREAD

**Issue by yangky11** [baldurk/renderdoc#362]
Hi,

I'm using renderdoc to inject GTA5. Since my game is on steam, I followed [this instruction](https://renderdoc.org/docs/window/capture_log_attach.html#global-process-hook) to do global process hooking.

After the game is launched, the in-app overlay appears and I can press F12 to capture. Then, I disabled global process hook as instructed. However, I cannot find the running instance under localhost in the 'attach to running instances' menu, and the diagnostic log shows that

```
RENDERDOC:  [14:26:46]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38920
RENDERDOC:  [14:26:46]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38921
RENDERDOC:  [14:26:46]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38922
RENDERDOC:  [14:26:46]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38923
RENDERDOC:  [14:26:47]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38924
RENDERDOC:  [14:26:47]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38925
RENDERDOC:  [14:26:47]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38926
RENDERDOC:  [14:26:48]    win32_network.cpp( 351) - Warning - Failed to connect to localhost:38927

```

Any idea? thanks.

**Comment by baldurk** [comment]
It's possible that enough sockets were opened that the final target process wasn't able to open a socket within that range, although that seems quite unlikely.

Do you have any firewall software running (windows built-in or 3rd party)? Perhaps there's some issue connecting because of that. When you've disabled the global hook, the UI is still running elevated as an administrator. You could try to re-launch the UI as a normal user and see if you have more luck.

**Comment by yangky11** [comment]
I just tried turn off windows built-in firewalls and launch the UI as a normal user but it didn't work.

**Comment by baldurk** [comment]
Can you find the logfile from the GTA5 process? There should be a `RenderDoc_app_<date>_<time>.log` in your temporary folder, it will be one of those and it will say the executable it's injecting into for each .log. I wonder if it has any messages about failing to bind or listen.

I added a couple of extra log lines that might be helpful. If you're able to build a copy of renderdoc locally, can you try with the code I just pushed? Otherwise it will be in the next nightly build.

Also worst case, if you have successfully injected and captured that's the hard part over - you just have to look in your temporary folder to find them and open them manually. Attaching is convenient and makes it work nicer (and I'd like to fix the problem!) but it's not required.

**Comment by yangky11** [comment]
Hi,

This is what I get using 2016-09-18 nightly build

RenderDoc_app_2016.09.17_21.12.35.log
```
RENDERDOC:  [21:12:35]             core.cpp( 312) - Log     - RenderDoc v0.32 x64 (b79f170e24f1c9b83b5addc4b080301672088860) capturing application
RENDERDOC:  [21:12:35]   win32_libentry.cpp(  66) - Log     - Loading into C:\Program Files\Rockstar Games\Grand Theft Auto V\GTA5.exe
RENDERDOC:  [21:12:35]            hooks.cpp(  49) - Log     - Loaded and hooked into kernel32.dll, PID 2476
RENDERDOC:  [21:12:35]            hooks.cpp(  49) - Log     - Loaded and hooked into opengl32.dll, PID 2476
RENDERDOC:  [21:12:35]            hooks.cpp(  49) - Log     - Loaded and hooked into d3d11.dll, PID 2476
RENDERDOC:  [21:12:35]            hooks.cpp(  49) - Log     - Loaded and hooked into d3d9.dll, PID 2476
RENDERDOC:  [21:12:35]            hooks.cpp(  49) - Log     - Loaded and hooked into dxgi.dll, PID 2476
RENDERDOC:  [21:12:35]            hooks.cpp(  49) - Log     - Loaded and hooked into d3d12.dll, PID 2476
RENDERDOC:  [21:12:35]            hooks.cpp(  49) - Log     - Loaded and hooked into vulkan-1.dll, PID 2476
RENDERDOC:  [21:12:42]     dxgi_wrapped.cpp( 518) - Warning - Querying IDXGIObject for interface: GUID {7abb6563-02bc-47c4-8ef9-acc4795edbcf}
RENDERDOC:  [21:12:43]    win32_network.cpp(  98) - Warning - accept: 10038
RENDERDOC:  [21:12:43]      d3d11_hooks.cpp( 267) - Log     - Created wrapped D3D11 device.
RENDERDOC:  [21:12:43]   target_control.cpp( 262) - Error   - Error in accept - shutting down server
RENDERDOC:  [21:12:43]     dxgi_wrapped.cpp( 518) - Warning - Querying IDXGIObject for interface: GUID {7abb6563-02bc-47c4-8ef9-acc4795edbcf}
RENDERDOC:  [21:12:43]      d3d11_hooks.cpp( 267) - Log     - Created wrapped D3D11 device.
RENDERDOC:  [21:12:43]     dxgi_wrapped.cpp( 518) - Warning - Querying IDXGIObject for interface: GUID {7abb6563-02bc-47c4-8ef9-acc4795edbcf}
RENDERDOC:  [21:12:43]     dxgi_wrapped.cpp( 518) - Warning - Querying IDXGIObject for interface: GUID {7abb6563-02bc-47c4-8ef9-acc4795edbcf}
RENDERDOC:  [21:15:29]     d3d11_device.cpp(2649) - Log     - Starting capture, frame 3274
RENDERDOC:  [21:15:29]     d3d11_device.cpp(2686) - Log     - Finished capture, Frame 3275
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5428
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5434
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5440
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5443
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5455
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5474
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5783
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6345
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6348
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6351
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6354
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6384
RENDERDOC:  [21:15:29]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:29]       serialiser.cpp(1447) - Log     - Compressed frame capture data from 59188672 to 38204179
RENDERDOC:  [21:15:29]       serialiser.cpp(1305) - Log     - Timer File writing - 200.905 ms
RENDERDOC:  [21:15:29]             core.cpp( 917) - Log     - Written to disk: C:\Users\yangky\AppData\Local\Temp\GTA5_2016.09.17_21.11.02_frame3275.rdc
RENDERDOC:  [21:15:32]     d3d11_device.cpp(2649) - Log     - Starting capture, frame 3326
RENDERDOC:  [21:15:32]     d3d11_device.cpp(2686) - Log     - Finished capture, Frame 3327
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5428
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5434
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5440
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5443
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5452
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5455
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5484
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 5783
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6345
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6348
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6351
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6354
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6384
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6485
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 6870
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 11183
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 11186
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1705) - Warning - Not serialising texture 2D initial state. ID 11189
RENDERDOC:  [21:15:32]     d3d11_device.cpp(1710) - Warning - Detected Write before Read of this target - assuming initial contents are unneeded.
Capture again with Save All Initials if this is wrong
RENDERDOC:  [21:15:32]       serialiser.cpp(1447) - Log     - Compressed frame capture data from 64325568 to 37600808
RENDERDOC:  [21:15:32]       serialiser.cpp(1305) - Log     - Timer File writing - 165.891 ms
RENDERDOC:  [21:15:32]             core.cpp( 917) - Log     - Written to disk: C:\Users\yangky\AppData\Local\Temp\GTA5_2016.09.17_21.11.02_frame3327.rdc
```

RenderDoc_replay_2016.09.17_21.16.01.log
```
RENDERDOC:  [21:16:01]             core.cpp( 309) - Log     - RenderDoc v0.32 x64 (b79f170e24f1c9b83b5addc4b080301672088860) loaded in replay application
RENDERDOC:  [21:16:05]  replay_renderer.cpp(1532) - Log     - Creating replay device for C:\Users\yangky\AppData\Local\Temp\GTA5_2016.09.17_21.11.02_frame3275.rdc
RENDERDOC:  [21:16:05]      d3d11_hooks.cpp( 267) - Log     - Created wrapped D3D11 device.
RENDERDOC:  [21:16:05]     d3d11_replay.cpp(2064) - Log     - Created device.
RENDERDOC:  [21:16:05]  replay_renderer.cpp(1551) - Log     - Created replay driver.
RENDERDOC:  [21:16:06]     d3d11_device.cpp(1090) - Log     - Timer chunk initialisation - 568.749 ms
RENDERDOC:  [21:16:10]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38920
RENDERDOC:  [21:16:11]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38921
RENDERDOC:  [21:16:11]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38922
RENDERDOC:  [21:16:11]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38923
RENDERDOC:  [21:16:11]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38924
RENDERDOC:  [21:16:12]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38925
RENDERDOC:  [21:16:12]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38926
RENDERDOC:  [21:16:12]    win32_network.cpp( 352) - Warning - Failed to connect to localhost:38927
```

**Comment by baldurk** [comment]
The relevant part seems to be this:

```
RENDERDOC:  [21:12:43]    win32_network.cpp(  98) - Warning - accept: 10038
RENDERDOC:  [21:12:43]   target_control.cpp( 262) - Error   - Error in accept - shutting down server
```

Which happens a 8 seconds after startup. If these are definitely the two logs though, this means the error happened long before you tried to connect in the UI - is that correct? That you started the game, made a couple of captures, and only tried to connect 4 minutes later? 8 seconds is long enough that it could be some kind of timeout, so this might be the first `accept()` call that is failing. I'm far from a networking expert so I'm just guessing.

It indicates something is going very wrong with the socket connection but I'm at a loss to say what exactly. Especially 10038 which is `WSAENOTSOCK` when clearly the socket should be fine. My instinct says this points to something else on the system interfering since it's so consistent for you but to my knowledge hasn't happened to anyone else, but it might be some weird edge case.

Can you test by launching a smaller test application normally from RenderDoc (without the global hook), and see if you can attach to it successfully? I'm curious if the problem is limited to the global hook and GTA5, or if it's system-wide for you.

Are you able to build RenderDoc yourself? The next step is to add a lot more debug logging, so I'll put it on a branch which won't go into the nightlies - either you'll need to build a local copy, or I can build one and upload it somewhere.

**Comment by yangky11** [comment]
Hi,

I tested with the Tutorial 1 in [Direct3D Samples](https://code.msdn.microsoft.com/windowsdesktop/Direct3D-Tutorial-Win32-829979ef), it worked fine. The problem seems to be specific for global hooks. I have no problem building RendorDoc

**Comment by baldurk** [comment]
I've pushed up to a branch debug-logging, can you compile from that branch and see what log output you get for the GTA5 process? It should at least give some clue exactly where and when the problem happens, hopefully that will lead somewhere else.

Please take careful note of the logfiles so that you can match the exact ones for both the captured process and the UI together, to get accurate timestamps from both sides of the socket.

**Comment by yangky11** [comment]
Hi,
My work requires some in-depth usages of RenderDoc, I'm going to read the code and hopefully will be able to figure out this problem myself. Thanks for your assistance!

**Comment by baldurk** [comment]
Sure thing, let me know if you need any more help.

**Comment by baldurk** [comment]
I'm not sure why I wasn't able to earlier, but I was able to actually repro this with GTA 5. It turned out to be a GTA 5 specific bug, the application was calling `WSACleanup()` before it had ever called `WSAStartup()` - which meant that it destroyed the sockets I had created while hooking in initially.

I fixed it just by hooking those two functions and preventing broken applications from doing anything - it just returns an error value without actually doing the cleanup. Hopefully that should fix it now.
